csharp/src/SeedLang/Runtime/NativeFunctions.cs

```diff
@@ -17,73 +17,115 @@
+    // Evaluates a value when it's not none. It's used by SeedCalc and interactive mode SeedX
+    // primarily.
+    private static Value EvalFunc(Value[] args, int offset, int length, VisualizerCenter vc,
+                                  Range range) {
+      if (!vc.EvalPublisher.IsEmpty()) {
+        if (length != 1) {
           throw new DiagnosticException(SystemReporters.SeedRuntime, Severity.Fatal, "", null,
                                         Message.RuntimeErrorIncorrectArgsCount);
         }
-        if (length == 0) {
-          return new Value(new List<Value>());
-        }
-        if (args[offset].IsList) {
-          return args[offset];
-        }
-        var list = new List<Value>();
-        for (int i = 0; i < args[offset].Length; i++) {
-          list.Add(args[offset][i]);
+        if (!args[offset].IsNone) {
+          vc.EvalPublisher.Notify(new EvalEvent(new ValueWrapper(args[offset]), range));
         }
-        return new Value(list);
-      }),
+      }
+      return new Value();
+    }
 
-      new NativeFunction(Range, (Value[] args, int offset, int length) => {
-        if (length == 1) {
-          return new Value(new Range((int)args[offset].AsNumber()));
-        } else if (length == 2) {
-          var range = new Range((int)args[offset].AsNumber(),(int)args[offset + 1].AsNumber());
-          return new Value(range);
-        } else if (length == 3) {
-          return new Value(new Range((int)args[offset].AsNumber(),
-                                     (int)args[offset + 1].AsNumber(),
-                                     (int)args[offset + 2].AsNumber()));
-        }
+    private static Value LenFunc(Value[] args, int offset, int length, VisualizerCenter vc,
+                                 Range callRange) {
+      if (length != 1) {
         throw new DiagnosticException(SystemReporters.SeedRuntime, Severity.Fatal, "", null,
                                       Message.RuntimeErrorIncorrectArgsCount);
-      }),
-    };
+      }
+      return new Value(args[offset].Length);
+    }
+
+    // Creates an empty list if the length of arguments is empty, and a list if the argument is a
+    // subscriptable value.
+    private static Value ListFunc(Value[] args, int offset, int length, VisualizerCenter vc,
+                                  Range callRange) {
+      if (length < 0 || length > 1) {
+        throw new DiagnosticException(SystemReporters.SeedRuntime, Severity.Fatal, "", null,
+                                      Message.RuntimeErrorIncorrectArgsCount);
+      }
+      if (length == 0) {
+        return new Value(new List<Value>());
+      }
+      if (args[offset].IsList) {
+        return args[offset];
+      }
+      var list = new List<Value>();
+      for (int i = 0; i < args[offset].Length; i++) {
+        list.Add(args[offset][i]);
+      }
+      return new Value(list);
+    }
+
+    private static Value PrintFunc(Value[] args, int offset, int length, VisualizerCenter vc,
```

Comment:
I think `print()` is not a function that does not do anything but notifies registered visualizers about the parameters.

A typical semantic meaning of `print` should be print contents to the standard output or the debug console.

Are we going to implement the real feature of `print`? Are we going to need something like Python's `__str__` as well to support `print`?

Reply:
The reason that I implement `print` as a notification to the client is SeedLang core doesn't know how to print. It might be calling `Console.Write` in SeedLang.Shell, or output to a Unity 3D log widget in the Sorting App. Then notification can be used to separate SeedLang core and the client.
Another approach could be just writing to the stdout, and let the client capture the content from the stdout. It might be a little difficult to handle.
@wixette Any thoughts about these two options?

__str__ can be implement later as an attribute of string value type. Current Value.AsString() method can be used to support `print`.

Reply:
print() is so common in script languages thus it is worth supporting it in different environments. A good and similar example is JavaScript's `console.log()`: In node.js commandline env, `console.log()` outputs to stdout since the console in the context is bound to stdout; while in a browser's window, `console.log()` outputs to browser's developer tools window, since the console object in this context is bound to the browser window.

I guess we need to support this binding mechanism for Python's `print` and  Lua's `print` too.

1. We will probably need a built-in library object `sys.stdout` (in SeedPython) or `io.stdout` (in SeedLua), that is bound to stdout by default. SeedLang's users can have an interface provided by SeedLang runtime to re-bind `sys.stdout` or `io.stdout` to another object. In this way, the client can easily customize the behavior of the `print` function.

2. If we do not plan to introduce `sys.stdout` and `io.stdout` for now, we can make the object hidden from the client code. E.g., to have an internal built-in object named `__stdout__`. Then, we provide an interface called `RedirectStdout(...)` in SeedLang's public interface.

3. Since we are building SeedLang in C#, the interface of `RedirectStdout(...)` could be something like below:

```
namespace SeedLang.Runtime {
  class ... {
    void RedirectStdout(System.IO.TextWriter writer, out System.IO.TextWriter oldWriter);
  }
}
```

In this way, users can implement their own TextWriter class to accept the text to be printed.